src/net.cpp

```diff
@@ -2642,17 +2651,140 @@ void CConnman::RecordBytesRecv(uint64_t bytes)
     nTotalBytesRecv += bytes;
 }
 
-void CConnman::RecordBytesRecvByMsgType(mapMsgTypeSize msgtype_bytes)
+void CConnman::RecordMsgStatsRecv(std::map<std::string, std::tuple<int, uint64_t>> msgtype_countbytes,
+                                  ConnectionType conn_type, Network net_type)
 {
-    for (auto const& msg_type_bytes : msgtype_bytes) {
-        auto i = m_msgtype_bytes_recv.find(msg_type_bytes.first);
+    for (auto const& msgtype_stats : msgtype_countbytes) {
+        std::string msg_type = msgtype_stats.first;
+        auto i = m_msgtype_bytes_recv.find(msg_type);
         if (i == m_msgtype_bytes_recv.end()) {
             i = m_msgtype_bytes_recv.find(NET_MESSAGE_TYPE_OTHER);
         }
 
         assert(i != m_msgtype_bytes_recv.end());
-        i->second += msg_type_bytes.second;
+        std::tuple<int, uint64_t> count_bytes = msgtype_stats.second;
+
+        int msg_count = std::get<0>(count_bytes);
+        uint64_t num_bytes = std::get<1>(count_bytes);
+
+        if (num_bytes > 0) {
+            i->second += num_bytes;
+
+            NetMsgStatsKey stats_key = {msg_type, conn_type, net_type};
+
+            // LogPrint(BCLog::NET, "\nstacie - incremeting received message count by %d, num_bytes by %d for (%s, %s, %s)\n",
+            //          msg_count, num_bytes, msg_type, ConnectionTypeAsString(conn_type), NetworkAsString(net_type));
+            auto stats_iterator = m_netmsg_stats_recv.find(stats_key);
+
+            // If this stat does not exist, add it (could also initialize all stats to zero?)
+            if (stats_iterator == m_netmsg_stats_recv.end()) {
+                NetMsgStatsValue stats_value = {msg_count, num_bytes};
+                m_netmsg_stats_recv.insert(std::map<NetMsgStatsKey, NetMsgStatsValue>::value_type(stats_key, stats_value));
+            } else {
+                // otherwise, update the message count and number of bytes
+                stats_iterator->second.msg_count += msg_count;
+                stats_iterator->second.num_bytes += num_bytes;
+            }
+        }
+    }
+}
+
+std::map<std::string, CConnman::NetMsgStatsValue> CConnman::AggregateNetMsgStats(const std::map<NetMsgStatsKey, NetMsgStatsValue>& raw_stats, std::vector<int> filters) const
```

Comment:
I don't think this function works correctly. The way I read it, it is going to sum all stats, unconditionally, under a single key of `aggregate_stats`. Anyway, I think it is better to do the aggregation in the RPC code and to return unaggregated stats from `connman` - there you can create directly the output JSON.

I would suggest to leave aggregation alone for now and come up with a completely working and polished code. Aggregation can even be done in a separate followup PR, after the first one which returns unaggregated data is merged.

Reply:
As of https://github.com/satsie/bitcoin/pull/3/commits/f5a5d46e0f0aa64a374d817a9e524fcdce792568 aggregation has been moved out of CConnman and into the RPC code. I felt like I was really close with it, so I kept on it, though it did take several revisions :sweat_smile: It is now working and returning a proper JSON object! (not all stats compressed into a single layer of an object)